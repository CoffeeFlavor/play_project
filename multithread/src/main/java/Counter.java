/**
 * @author : jennie
 * @date: 2018/4/27
 * @Time: 14:05
 */
public class Counter {

    long count=0;

    public synchronized void add(long  value){
        this.count=this.count+value;
    }

    public  class  CounterThread extends Thread{
        protected Counter counter=null;

        public CounterThread(Counter counter) {
            this.counter = counter;
        }

        @Override
        public void run() {
            for (int i = 0; i < 10; i++) {
                counter.add(i);
            }
        }
    }
}
