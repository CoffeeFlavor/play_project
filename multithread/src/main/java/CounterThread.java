/**
 * @author : jennie
 * @date: 2018/5/4
 * @Time: 10:33
 */
public class CounterThread extends Thread{

    protected Counter counter=null;

    public CounterThread(Counter counter) {
        this.counter = counter;
    }

    @Override
    public void run() {
        for (int i = 0; i < 10; i++) {
            counter.add(i);
        }
    }
}
