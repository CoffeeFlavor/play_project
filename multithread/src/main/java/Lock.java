/**
 * @author : jennie
 * @date: 2018/5/7
 * @Time: 18:23
 */
public class Lock {

    private Boolean isLocked=false;

    private Thread lockingThread=null;

    private synchronized void lock() throws InterruptedException{
        while (isLocked){
            wait();
        }
        isLocked=true;
        lockingThread=Thread.currentThread();
    }

    public synchronized void unLock(){
       if (this.lockingThread!=Thread.currentThread()){
           throw new IllegalMonitorStateException("" +
                   "calling thread has not locked this lock ");
       }
       isLocked=false;
       lockingThread=null;
       notify();
    }
}
