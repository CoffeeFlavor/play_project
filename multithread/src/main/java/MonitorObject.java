/**
 * @author : jennie
 * @date: 2018/5/4
 * @Time: 11:54
 */
public class MonitorObject {
}
