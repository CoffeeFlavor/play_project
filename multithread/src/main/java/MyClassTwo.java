/**
 * @author : jennie
 * @date: 2018/5/2
 * @Time: 14:56
 */
public class MyClassTwo {

     synchronized void  test4(){
        synchronized (this){
            try {
                Thread.sleep(10000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("test4");
        }

    }

    void test5(){

        synchronized (this) {
            System.out.println("test5");
        }
    }
}
