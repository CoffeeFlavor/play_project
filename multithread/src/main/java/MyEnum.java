/**
 * @author : jennie
 * @date: 2018/5/9
 * @Time: 15:15
 */
public enum  MyEnum {
    AB,
    CD;
   private Boolean hasNotice=Boolean.FALSE;


    public void doWait() {
        synchronized (this) {
            if (!hasNotice) {
                try {
                    this.wait();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
            //获取接收数据
            hasNotice = Boolean.FALSE;
        }
    }

    public void doNotify() {
        synchronized (this) {
            //有数据需要处理,唤醒线程
            hasNotice = Boolean.TRUE;
            this.notify();
        }
    }
}
