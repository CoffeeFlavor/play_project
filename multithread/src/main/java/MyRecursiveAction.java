import java.util.AbstractList;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.RecursiveAction;

/**
 * @author : jennie
 * @date: 2018/4/26
 * @Time: 18:07
 */
public class MyRecursiveAction extends RecursiveAction {

    private long workLoad=0;

    public MyRecursiveAction(long workLoad) {
        this.workLoad = workLoad;
    }

    @Override
    protected void compute() {
        if (this.workLoad>16){
            System.out.println("Splitting workLoad:"+this.workLoad);

            List<MyRecursiveAction> subTasks=new ArrayList<>();

            subTasks.addAll(creatSubtasks());
            for (MyRecursiveAction subTask : subTasks) {
                subTask.fork();
            }
        } else {
            System.out.println("Doing workLoad myself: "+this.workLoad);
        }

    }

    private List<MyRecursiveAction> creatSubtasks(){
        List<MyRecursiveAction> subTasks=new ArrayList<>();
        MyRecursiveAction subTasks1=new MyRecursiveAction(this.workLoad/2);
        MyRecursiveAction subTasks2=new MyRecursiveAction(this.workLoad/3);

        subTasks.add(subTasks1);
        subTasks.add(subTasks2);
        return subTasks;
    }
}
