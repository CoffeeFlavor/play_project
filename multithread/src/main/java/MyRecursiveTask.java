import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.RecursiveTask;

/**
 * @author : jennie
 * @date: 2018/4/27
 * @Time: 11:24
 */
public class MyRecursiveTask extends RecursiveTask<Long>{

    private long workLoad=0;

    public MyRecursiveTask(long workLoad) {
        this.workLoad = workLoad;
    }

    @Override
    protected Long compute() {
        if (this.workLoad>16){
            List<MyRecursiveTask> subTasks=new ArrayList<>();
            subTasks.addAll(createSubTask());
            for (MyRecursiveTask subTask : subTasks) {
                subTask.fork();
            }
            long result=0;
            for (MyRecursiveTask subTask : subTasks) {
                result+=subTask.join();
            }
            return result;
        }else {

            System.out.println("Doing workLoad myself: " +this.workLoad);
            return this.workLoad*3;
        }
    }

    private List<MyRecursiveTask> createSubTask(){
        List<MyRecursiveTask> subTasks=new ArrayList<>();
        MyRecursiveTask subTask1=new MyRecursiveTask(workLoad/2);
        MyRecursiveTask subTask2=new MyRecursiveTask(workLoad/2);
        subTasks.add(subTask1);
        subTasks.add(subTask2);
        return subTasks;
    }
}
