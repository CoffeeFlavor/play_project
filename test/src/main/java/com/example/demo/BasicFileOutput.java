package com.example.demo;

import java.io.*;

/**
 * User: jennie
 * Date: 2017/10/10
 * Time: 15:25
 */
public class BasicFileOutput {
    static String file = "F:\\play_project\\test\\src\\main\\java\\com\\example\\demo\\BasicFileOutput.out";

    public static void main(String[] args) {
        try {
            BufferedReader in = new BufferedReader(new StringReader(BufferedInputFile.read("F:\\play_project\\test\\src\\main\\java\\com\\example\\demo\\BasicFileOutput.java")));
            PrintWriter out = new PrintWriter(file);
            int lineCount = 1;
            String s;
            while ((s = in.readLine()) != null)
                out.println(lineCount++ + ": " + s);
            out.close();
            System.out.println(BufferedInputFile.read(file));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
