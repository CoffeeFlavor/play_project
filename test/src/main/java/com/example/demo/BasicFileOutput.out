1: package com.example.demo;
2: 
3: import java.io.*;
4: 
5: /**
6:  * User: jennie
7:  * Date: 2017/10/10
8:  * Time: 15:25
9:  */
10: public class BasicFileOutput {
11:     static String file = "F:\\play_project\\test\\src\\main\\java\\com\\example\\demo\\BasicFileOutput.out";
12: 
13:     public static void main(String[] args) {
14:         try {
15:             BufferedReader in = new BufferedReader(new StringReader(BufferedInputFile.read("F:\\play_project\\test\\src\\main\\java\\com\\example\\demo\\BasicFileOutput.java")));
16:             PrintWriter out = new PrintWriter(file);
17:             int lineCount = 1;
18:             String s;
19:             while ((s = in.readLine()) != null)
20:                 out.println(lineCount++ + ": " + s);
21:             out.close();
22:             System.out.println(BufferedInputFile.read(file));
23:         } catch (IOException e) {
24:             e.printStackTrace();
25:         }
26:     }
27: }
